tests/integration/dhcp_test.py

```diff
@@ -0,0 +1,154 @@
+# the Free Software Foundation, either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+#
+import os
+import pytest
+
+from libnmstate import netapplier
+from libnmstate import nm
+
+from .testlib import assertlib
+from .testlib import statelib
+from .testlib.statelib import INTERFACES
+from .testlib import cmd as libcmd
+
+
+IPV4_ADDRESS1 = '192.0.2.251'
+IPV4_ADDRESS2 = '192.0.2.252'
+IPV6_ADDRESS1 = '2001:db8:1::1'
+IPV6_ADDRESS2 = '2001:db8:2::1'
+
+DHCP_SRV_NIC = 'dhcpsrv'
+DHCP_CLI_NIC = 'dhcpcli'
+
+DNSMASQ_CONF_STR = """
+interface={}
+dhcp-range=192.0.2.200,192.0.2.250,255.255.255.0,48h
+enable-ra
+dhcp-range=2001:db8:1::100,2001:db8:1::fff,64,480h
+""".format(DHCP_SRV_NIC)
+
+DNSMASQ_CONF_PATH = '/etc/dnsmasq.d/nmstate.conf'
+
+SYSFS_DISABLE_IPV6_FILE = '/proc/sys/net/ipv6/conf/{}/disable_ipv6'.format(
+    DHCP_SRV_NIC)
+
+# Python 2 does not have FileNotFoundError and treat file not exist as IOError
+try:
+    FileNotFoundError
+except NameError:
+    FileNotFoundError = IOError
+
+
+@pytest.fixture(scope='module')
+def start_dhcpd(request):
+    def fin():
+        libcmd.exec_cmd(['systemctl', 'stop', 'dnsmasq'])
+        libcmd.exec_cmd(['ip', 'link', 'del', 'dev', DHCP_SRV_NIC])
+        try:
+            os.unlink(DNSMASQ_CONF_PATH)
+        except FileNotFoundError:
+            pass
+
+    request.addfinalizer(fin)
+
+    assert libcmd.exec_cmd(['ip', 'link', 'add', DHCP_SRV_NIC, 'type', 'veth',
+                            'peer', 'name', DHCP_CLI_NIC])[0] == 0
+    fd = open(DNSMASQ_CONF_PATH, 'w')
+    fd.write(DNSMASQ_CONF_STR)
+    fd.close()
+    assert libcmd.exec_cmd(['ip', 'link', 'set', DHCP_SRV_NIC, 'up'])[0] == 0
+    assert libcmd.exec_cmd(['ip', 'link', 'set', DHCP_CLI_NIC, 'up'])[0] == 0
+    assert libcmd.exec_cmd(['ip', 'addr', 'add', IPV4_ADDRESS1, 'dev',
+                            DHCP_SRV_NIC])[0] == 0
+    fd = open(SYSFS_DISABLE_IPV6_FILE, 'w')
+    fd.write('0')
+    fd.close()
+    assert libcmd.exec_cmd(['ip', 'addr', 'add', IPV6_ADDRESS1, 'dev',
+                            DHCP_SRV_NIC])[0] == 0
+    assert libcmd.exec_cmd(['systemctl', 'restart', 'dnsmasq'])[0] == 0
+
+
+def test_ipv4_dhcp(start_dhcpd):
+    desired_state = statelib.show_only((DHCP_CLI_NIC,))
+    dhcp_cli_desired_state = desired_state[INTERFACES][0]
+    dhcp_cli_desired_state['state'] = 'up'
+    dhcp_cli_desired_state['ipv4']['enabled'] = True
+    dhcp_cli_desired_state['ipv4']['dhcp'] = True
+    dhcp_cli_desired_state['ipv6']['enabled'] = True
+    netapplier.apply(desired_state)
```

Comment:
Please separate using a newline between the setup preparation, execution and assertion/check.